diffsims/sims/diffraction_simulation.py

```diff
@@ -44,35 +45,103 @@ class DiffractionSimulation:
+            and coordinates.ndim == indices.ndim == 2
+            and intensities.ndim == 1
+        ):
+            self._coordinates = coordinates
+            self._indices = indices
+            self._intensities = intensities
+        else:
+            raise ValueError(
+                "Coordinate, intensity, and indices lists must be of the correct and matching shape."
+            )
         self.calibration = calibration
         self.offset = offset
         self.with_direct_beam = with_direct_beam
 
+    def __len__(self):
+        return self.coordinates.shape[0]
+
+    def __getitem__(self, sliced):
+        coords = self.coordinates[sliced]
+        inds = self.indices[sliced]
+        ints = self.intensities[sliced]
+        if coords.ndim == 1:
+            coords = coords[None, :]
+            inds = inds[None, :]
+            ints = ints[None]
+        # some valid numpy slices will create invalid shapes for diffraction simulation
+        if coords.ndim > 2 or coords.shape[1] > 3 or coords.shape[1] < 2:
+            raise ValueError(f"Invalid slice: {sliced}")
+        return DiffractionSimulation(
+            coords,
+            indices=inds,
+            intensities=ints,
+            calibration=self.calibration,
+            offset=self.offset,
+            with_direct_beam=self.with_direct_beam,
+        )
+
+    def _combine_data(self, other):
+        coords = np.concatenate([self._coordinates, other._coordinates], axis=0)
+        inds = np.concatenate([self._indices, other._indices], axis=0)
+        ints = np.concatenate([self._intensities, other._intensities], axis=0)
+        return coords, inds, ints
+
+    def __add__(self, other):
+        coords, inds, ints = self._combine_data(other)
+        return DiffractionSimulation(
+            coords,
+            indices=inds,
+            intensities=ints,
+            calibration=self.calibration,
+            offset=self.offset,
+            with_direct_beam=self.with_direct_beam,
+        )
+
+    def extend(self, other):
+        """Add the diffraction spots from another DiffractionSimulation"""
+        coords, inds, ints = self._combine_data(other)
+        self._coordinates = coords
+        self._indices = inds
+        self._ints = ints
+
+    @property
+    def indices(self):
+        return self._indices[self.direct_beam_mask]
+
+    @indices.setter
+    def indices(self, indices):
+        self._indices[self.direct_beam_mask] = indices
+
     @property
     def calibrated_coordinates(self):
         """ndarray : Coordinates converted into pixel space."""
-        coordinates = np.copy(self.coordinates)
-        coordinates[:, 0] += self.offset[0]
-        coordinates[:, 1] += self.offset[1]
-        coordinates[:, 0] /= self.calibration[0]
-        coordinates[:, 1] /= self.calibration[1]
-        return coordinates
+        if self.calibration is not None:
+            return (self.coordinates[:, :2] + np.array(self.offset)) / np.array(
```

Comment:
I think most people pass in arguments as tuples or lists and expect it to work. Since offset is only used in this method, I think explicitly converting to an array to allow broadcasting makes sense here. But I might move it to the init method.

Reply:
I also applied this to the calibration, it is automatically converted to a numpy array.